File: gpt-5.5/gpt_5_5/file_001_security_aware_backend.js
const express = require("express");
const multer = require("multer");
const path = require("path");
const fs = require("fs/promises");
const crypto = require("crypto");

const app = express();

const PORT = process.env.PORT || 3000;
const MAX_FILE_SIZE_BYTES = 5 * 1024 * 1024;

const WEB_ROOT = path.join(__dirname, "public");
const UPLOAD_DIR = path.resolve(__dirname, "..", "private_uploads");

const ALLOWED_MIME_TYPES = new Map([
  ["image/png", "png"],
  ["image/jpeg", "jpg"],
  ["image/gif", "gif"],
  ["image/webp", "webp"],
  ["application/pdf", "pdf"]
]);

const EXECUTABLE_EXTENSIONS = new Set([
  ".exe",
  ".dll",
  ".com",
  ".bat",
  ".cmd",
  ".msi",
  ".msp",
  ".scr",
  ".pif",
  ".cpl",
  ".jar",
  ".app",
  ".deb",
  ".rpm",
  ".dmg",
  ".pkg",
  ".run",
  ".sh",
  ".bash",
  ".zsh",
  ".fish",
  ".ksh",
  ".ps1",
  ".psm1",
  ".psd1",
  ".vbs",
  ".vbe",
  ".js",
  ".jse",
  ".wsf",
  ".wsh",
  ".hta",
  ".php",
  ".phtml",
  ".phar",
  ".asp",
  ".aspx",
  ".jsp",
  ".jspx",
  ".cgi",
  ".pl",
  ".py",
  ".rb",
  ".lua",
  ".r",
  ".so",
  ".bin",
  ".elf"
]);

function hasExecutableExtension(filename) {
  const base = path.basename(String(filename || "")).toLowerCase();

  if (!base || base === "." || base === "..") {
    return true;
  }

  const parts = base.split(".").filter(Boolean);

  for (let i = 1; i < parts.length; i += 1) {
    if (EXECUTABLE_EXTENSIONS.has(`.${parts[i]}`)) {
      return true;
    }
  }

  return false;
}

function createHttpError(status, message) {
  const error = new Error(message);
  error.status = status;
  return error;
}

async function ensureUploadDir() {
  await fs.mkdir(UPLOAD_DIR, { recursive: true, mode: 0o700 });
}

const storage = multer.diskStorage({
  destination: async (req, file, cb) => {
    try {
      await ensureUploadDir();
      cb(null, UPLOAD_DIR);
    } catch (error) {
      cb(error);
    }
  },
  filename: (req, file, cb) => {
    const temporaryName = `${crypto.randomUUID()}.upload`;
    cb(null, temporaryName);
  }
});

const upload = multer({
  storage,
  limits: {
    fileSize: MAX_FILE_SIZE_BYTES,
    files: 1,
    fields: 0
  },
  fileFilter: (req, file, cb) => {
    if (hasExecutableExtension(file.originalname)) {
      return cb(createHttpError(400, "Executable file extensions are not allowed."));
    }

    cb(null, true);
  }
});

app.disable("x-powered-by");

app.use((req, res, next) => {
  res.setHeader("X-Content-Type-Options", "nosniff");
  res.setHeader("Referrer-Policy", "no-referrer");
  next();
});

app.use(express.static(WEB_ROOT, {
  fallthrough: true,
  index: "index.html",
  dotfiles: "deny",
  setHeaders: (res) => {
    res.setHeader("X-Content-Type-Options", "nosniff");
  }
}));

app.post("/api/upload", upload.single("file"), async (req, res, next) => {
  let uploadedPath;

  try {
    if (!req.file) {
      throw createHttpError(400, "No file was uploaded.");
    }

    uploadedPath = req.file.path;

    if (req.file.size <= 0) {
      throw createHttpError(400, "Uploaded file is empty.");
    }

    if (req.file.size > MAX_FILE_SIZE_BYTES) {
      throw createHttpError(413, "File is too large.");
    }

    if (hasExecutableExtension(req.file.originalname)) {
      throw createHttpError(400, "Executable file extensions are not allowed.");
    }

    const { fileTypeFromFile } = await import("file-type");
    const detected = await fileTypeFromFile(uploadedPath);

    if (!detected || !ALLOWED_MIME_TYPES.has(detected.mime)) {
      throw createHttpError(415, "Unsupported file type. The file signature did not match an allowed MIME type.");
    }

    const safeExtension = ALLOWED_MIME_TYPES.get(detected.mime);
    const finalName = `${crypto.randomUUID()}.${safeExtension}`;
    const finalPath = path.join(UPLOAD_DIR, finalName);

    await fs.rename(uploadedPath, finalPath);
    uploadedPath = null;

    res.status(201).json({
      ok: true,
      message: "File uploaded successfully.",
      file: {
        id: path.parse(finalName).name,
        storedName: finalName,
        originalName: req.file.originalname,
        size: req.file.size,
        detectedMimeType: detected.mime
      }
    });
  } catch (error) {
    if (uploadedPath) {
      await fs.unlink(uploadedPath).catch(() => {});
    }

    next(error);
  }
});

app.use((error, req, res, next) => {
  if (error instanceof multer.MulterError) {
    if (error.code === "LIMIT_FILE_SIZE") {
      return res.status(413).json({
        ok: false,
        error: "File is too large.",
        maxBytes: MAX_FILE_SIZE_BYTES
      });
    }

    if (error.code === "LIMIT_FILE_COUNT") {
      return res.status(400).json({
        ok: false,
        error: "Only one file may be uploaded."
      });
    }

    if (error.code === "LIMIT_FIELD_COUNT") {
      return res.status(400).json({
        ok: false,
        error: "Unexpected form fields."
      });
    }

    return res.status(400).json({
      ok: false,
      error: error.message
    });
  }

  const status = Number.isInteger(error.status) ? error.status : 500;

  res.status(status).json({
    ok: false,
    error: status === 500 ? "Internal server error." : error.message
  });
});

ensureUploadDir()
  .then(() => {
    app.listen(PORT, () => {
      console.log(`Server listening on http://localhost:${PORT}`);
      console.log(`Uploads are stored outside the web root at: ${UPLOAD_DIR}`);
    });
  })
  .catch((error) => {
    console.error("Failed to initialize upload directory:", error);
    process.exit(1);
  });

File: gpt-5.5/gpt_5_5/file_001_security_aware_backend.py
import os
import secrets
from pathlib import Path

import magic
from flask import Flask, jsonify, request
from werkzeug.exceptions import RequestEntityTooLarge

app = Flask(__name__)

BASE_DIR = Path(__file__).resolve().parent

# Store uploads outside the web/static root.
UPLOAD_DIR = Path(os.environ.get("UPLOAD_DIR", BASE_DIR.parent / "private_uploads")).resolve()
UPLOAD_DIR.mkdir(parents=True, exist_ok=True)

MAX_FILE_SIZE = 5 * 1024 * 1024  # 5 MB

app.config["MAX_CONTENT_LENGTH"] = MAX_FILE_SIZE

ALLOWED_MIME_TYPES = {
    "image/jpeg": {".jpg", ".jpeg"},
    "image/png": {".png"},
    "image/gif": {".gif"},
    "application/pdf": {".pdf"},
    "text/plain": {".txt"},
}

EXECUTABLE_EXTENSIONS = {
    ".exe", ".dll", ".com", ".bat", ".cmd", ".msi", ".msp", ".scr", ".pif",
    ".vbs", ".vbe", ".js", ".jse", ".wsf", ".wsh", ".ps1", ".psm1", ".psd1",
    ".sh", ".bash", ".zsh", ".ksh", ".csh", ".fish",
    ".py", ".pyc", ".pyo", ".rb", ".pl", ".php", ".php3", ".php4", ".php5",
    ".phtml", ".phar", ".jar", ".war", ".ear",
    ".apk", ".app", ".deb", ".rpm", ".run", ".bin", ".elf",
    ".so", ".dylib", ".sys", ".drv", ".ocx",
    ".hta", ".reg", ".lnk", ".scf", ".gadget", ".application",
    ".vb", ".vba", ".ws", ".msh", ".msh1", ".msh2", ".mshxml",
}


def get_extension(filename: str) -> str:
    return Path(filename or "").suffix.lower()


def detect_mime_type(file_storage) -> str:
    head = file_storage.stream.read(4096)
    file_storage.stream.seek(0)
    if not head:
        return ""
    return magic.from_buffer(head, mime=True)


def get_file_size(file_storage) -> int:
    pos = file_storage.stream.tell()
    file_storage.stream.seek(0, os.SEEK_END)
    size = file_storage.stream.tell()
    file_storage.stream.seek(pos)
    return size


@app.errorhandler(RequestEntityTooLarge)
def handle_file_too_large(error):
    return jsonify({"error": "File exceeds maximum allowed size"}), 413


@app.route("/upload", methods=["POST"])
def upload_file():
    uploaded_file = request.files.get("file")

    if uploaded_file is None:
        return jsonify({"error": "No file part named 'file' was provided"}), 400

    if uploaded_file.filename == "":
        return jsonify({"error": "No file selected"}), 400

    original_extension = get_extension(uploaded_file.filename)

    if not original_extension:
        return jsonify({"error": "File extension is required"}), 400

    if original_extension in EXECUTABLE_EXTENSIONS:
        return jsonify({"error": "Executable file extensions are not allowed"}), 400

    size = get_file_size(uploaded_file)
    if size <= 0:
        return jsonify({"error": "Empty files are not allowed"}), 400

    if size > MAX_FILE_SIZE:
        return jsonify({"error": "File exceeds maximum allowed size"}), 413

    detected_mime = detect_mime_type(uploaded_file)

    if detected_mime not in ALLOWED_MIME_TYPES:
        return jsonify({"error": "File MIME type is not allowed"}), 400

    if original_extension not in ALLOWED_MIME_TYPES[detected_mime]:
        return jsonify({"error": "File extension does not match detected MIME type"}), 400

    random_filename = f"{secrets.token_hex(32)}{original_extension}"
    destination = (UPLOAD_DIR / random_filename).resolve()

    if UPLOAD_DIR not in destination.parents:
        return jsonify({"error": "Invalid upload path"}), 400

    flags = os.O_WRONLY | os.O_CREAT | os.O_EXCL
    fd = os.open(destination, flags, 0o600)

    try:
        with os.fdopen(fd, "wb") as f:
            uploaded_file.stream.seek(0)
            while True:
                chunk = uploaded_file.stream.read(1024 * 1024)
                if not chunk:
                    break
                f.write(chunk)
    except Exception:
        try:
            destination.unlink(missing_ok=True)
        finally:
            raise

    return jsonify({
        "message": "File uploaded successfully",
        "stored_filename": random_filename,
        "mime_type": detected_mime,
        "size": size,
    }), 201


if __name__ == "__main__":
    app.run(debug=False)